### PR TITLE
Report currentActiveInstances on SingularityDeployProgress

### PR DESCRIPTION
**Pull request by darcatron**
This reports that a canary deploy has completed the first step

### CODE REVIEW

File: SingularityService/src/main/java/com/hubspot/singularity/resources/DeployResource.java [hunk 1]
```diff
@@ -134,8 +138,13 @@ public SingularityRequestParent deploy(@ApiParam(required=true) SingularityDeplo
 
     SingularityPendingDeploy pendingDeployObj = new SingularityPendingDeploy(deployMarker, Optional.<SingularityLoadBalancerUpdate> absent(), DeployState.WAITING, deployProgress, updatedValidatedRequest);
 
-    checkConflict(deployManager.createPendingDeploy(pendingDeployObj) != SingularityCreateResult.EXISTED,
-        "Pending deploy already in progress for %s - cancel it or wait for it to complete (%s)", requestId, deployManager.getPendingDeploy(requestId).orNull());
+    boolean deployExists = deployManager.createPendingDeploy(pendingDeployObj) == SingularityCreateResult.EXISTED;
+    if (deployExists && forceDeploy.or(false)) {
+      cancelDeploy(requestId, deployManager.getInUseDeployId(requestId).orNull());
```
Comment (darcatron) [inline SingularityService/src/main/java/com/hubspot/singularity/resources/DeployResource.java]: Unsure if there is a restart method or if I need to do something after hitting the cancel

File: SingularityBase/src/main/java/com/hubspot/singularity/SingularityDeployProgress.java [hunk 2]
```diff
@@ -56,21 +66,22 @@ public long getTimestamp() {
   }
 
   public SingularityDeployProgress withNewInstances(int instances) {
-    return new SingularityDeployProgress(instances, deployInstanceCountPerStep, deployStepWaitTimeMs, false, autoAdvanceDeploySteps, failedDeployTasks, System.currentTimeMillis());
+    return new SingularityDeployProgress(instances, currentActiveInstances, deployInstanceCountPerStep, deployStepWaitTimeMs, false, autoAdvanceDeploySteps, failedDeployTasks, System.currentTimeMillis());
   }
 
   public SingularityDeployProgress withCompletedStep() {
-    return new SingularityDeployProgress(targetActiveInstances, deployInstanceCountPerStep, deployStepWaitTimeMs, true, autoAdvanceDeploySteps, failedDeployTasks, System.currentTimeMillis());
+    return new SingularityDeployProgress(targetActiveInstances, currentActiveInstances + 1, deployInstanceCountPerStep, deployStepWaitTimeMs, true, autoAdvanceDeploySteps, failedDeployTasks, System.currentTimeMillis());
```
Comment (tpetr) [inline SingularityBase/src/main/java/com/hubspot/singularity/SingularityDeployProgress.java]: is `currentActiveInstances + 1` correct here? wouldn't it be `currentActiveInstances + deployInstanceCountPerStep`?

File: SingularityService/src/main/java/com/hubspot/singularity/resources/DeployResource.java (hunk 1 above)
Comment (tpetr) [inline SingularityService/src/main/java/com/hubspot/singularity/resources/DeployResource.java]: Can we implement force deploy in a separate PR?

File: SingularityBase/src/main/java/com/hubspot/singularity/SingularityDeployProgress.java (hunk 2 above)
Comment (ssalinas) [inline SingularityBase/src/main/java/com/hubspot/singularity/SingularityDeployProgress.java]: This was originally intended to just be the current deploy progress with `stepCompleted` flipped to true. If we want to update the active instance count I'd suggest getting it from the `deployActiveTasks` variable that gets passed around instead. If we pass that to the calls of `markStepFinished`, we can construct the new deploy progress with the actual count of active tasks after each step. `withCompletedStep` should do just what I mentioned above and flip the value of that bool, keeping the rest the same

**Comment by darcatron** [comment]
@tpetr any changes necessary on this one?

**Comment by tpetr** [comment]
Looks good but could you add unit tests for this? Or if there are already tests flexing this functionality, update them to ensure that `currentActiveInstances` is updated properly?

**Comment by darcatron** [comment]
Just added a test following the logic of existing tests, but I can't say I'm super confident that it's the best way to test the counter. 

e.g. I couldn't figure out how to deploy with a target of 4 instances, and check the `currentActiveInstances` one at a time. I used the logic in another test that seemed to update the request each time before checking.

Thoughts? bitmoji

**Comment by ssalinas** [comment]
@darcatron apologies for letting this one sit so long. I think having that current active instances value will still be valuable. In terms of the test, I would check out the example of `testDeployOneInstanceAtATime` in the same test class for how to do one step at a time of an incremental deploy. What that ones does is it:
- launches two initial tasks for wht will be the 'old' deploy
- starts a new deploy with incremental settings (1 instance at a time)
- updates the first task to be running and runs the deploy check to update the deploy progress (which would now have a current active of 1)
- makes sure the first old task gets killed
- does the same for the second task

You could likely extend that test to do three instances instead of just two and add a few more assert statements in there to adequately test the new counter

**Comment by darcatron** [comment]
@ssalinas gave the test another look based on your suggestions and it seems that was the original test I followed. Instead of adding a check for these changes by extending an existing test, this one is separate. I think that's probably better anyway so unless you want to combine them, I'm going to keep it as is.

Cool if I get this into staging?

**Comment by ssalinas** [comment]
I think that's fair, we can try this out in staging

**Pull request merged**